File: typescript/src/astro/hand-authored-regions.ts
/**
 * Hand-authored named sectors, resolved by galactic position.
 *
 * The procedural generator only produces the boxel name for a position. Systems
 * that physically fall inside a hand-authored nebula or cluster region (Pleiades,
 * Coalsack, the NGC/IC/Col catalogues, …) are instead displayed by the game under
 * that region's name. Each region is one or more spheres in galactic light-years
 * (Sol at the origin); membership is first-match sphere containment.
 *
 * Sphere records are compiled and cross-checked against EDSM and Spansh; see
 * [`data/astro/SOURCES.md`](https://github.com/DarkSession/Elite-Dangerous-Almanac/blob/main/data/astro/SOURCES.md) for provenance and source terms.
 *
 * The list is pre-sorted **smallest-radius-first**, which reproduces the game's
 * overlap priority: when spheres overlap, the most specific (smallest) region
 * wins. Data is loaded once from shared JSON ([`data/astro/`](https://github.com/DarkSession/Elite-Dangerous-Almanac/tree/main/data/astro)).
 *
 * @remarks
 * "Region" is overloaded in this galaxy. A {@link HandAuthoredRegion} is a *hand-authored
 * named sector* (Pleiades, Coalsack, …) — not to be confused with a **procedural
 * sector** (`./sector-name`) or a **galactic codex region** (`./codex-region`,
 * the 42 codex zones). Use {@link findHandAuthoredRegionAt} here for the first, and
 * `findCodexRegionAt` from `./codex-region-lookup` for the third.
 *
 * @packageDocumentation
 */

import type { GalacticPosition } from './galactic-position.js';
import handAuthoredData from '../../../data/astro/hand-authored-regions.jsonc' with { type: 'json' };
import { deepFreeze } from '../internal/deep-freeze.js';

/**
 * One sphere of a hand-authored region (centre and radius in light-years).
 *
 * @example
 * ```ts
 * import type { HandAuthoredSphere } from '@elite-dangerous-almanac/core/astro/hand-authored-regions';
 *
 * const sphere: HandAuthoredSphere = { cx: -80.6, cy: -146.7, cz: -343.3, r: 100 };
 * ```
 */
export interface HandAuthoredSphere {
    /** Sphere centre X in light-years (Sol at origin). */
    readonly cx: number;
    /** Sphere centre Y in light-years. */
    readonly cy: number;
    /** Sphere centre Z in light-years. */
    readonly cz: number;
    /** Sphere radius in light-years. */
    readonly r: number;
}

/**
 * A hand-authored named sector: its canonical name and the spheres it occupies.
 *
 * @remarks
 * Whether the region needs a permit is not stored here — 28 of these regions are
 * permit-locked, and `isPermitLockedRegionName` in `./permit-locks` is the single
 * place that knows which. Pass {@link HandAuthoredRegion.name} to it.
 *
 * @example
 * ```ts
 * import { HAND_AUTHORED_REGIONS } from '@elite-dangerous-almanac/core/astro/hand-authored-regions';
 *
 * const pleiades = HAND_AUTHORED_REGIONS.find((region) => region.name === 'Pleiades Sector');
 * ```
 */
export interface HandAuthoredRegion {
    /** Canonically-cased region name, e.g. `Pleiades Sector`. */
    readonly name: string;
    /** The spheres whose union defines the region's volume. */
    readonly spheres: readonly HandAuthoredSphere[];
}

/**
 * All hand-authored regions, sorted smallest-radius-first (overlap priority).
 *
 * @example
 * ```ts
 * import { HAND_AUTHORED_REGIONS } from '@elite-dangerous-almanac/core/astro/hand-authored-regions';
 *
 * HAND_AUTHORED_REGIONS[0]?.name;
 * ```
 */
export const HAND_AUTHORED_REGIONS: readonly HandAuthoredRegion[] = deepFreeze(
    handAuthoredData as readonly HandAuthoredRegion[],
);

/**
 * The hand-authored region containing a galactic point, or `null` for procedural
 * space.
 *
 * Because {@link HAND_AUTHORED_REGIONS} is sorted smallest-radius-first, the first region
 * whose sphere set contains the point is the most specific one — matching the
 * game's overlap priority.
 *
 * @remarks
 * This resolves a *hand-authored named sector* only. For the codex region a point
 * falls in, use `findCodexRegionAt` from `./codex-region-lookup` instead.
 *
 * @param position - Galactic position in light-years (Sol at origin). All three axes
 * are used — hand-authored regions are 3-D spheres, so `y` matters here (unlike the
 * flat `findCodexRegionAt`). A {@link ProceduralSystem.position} value can be passed directly.
 * @returns The containing region, or `null` if the point is in procedural space.
 * @example
 * ```ts
 * import { findHandAuthoredRegionAt } from '@elite-dangerous-almanac/core/astro/hand-authored-regions';
 *
 * findHandAuthoredRegionAt({ x: -80.6, y: -146.7, z: -343.3 })?.name;
 * // -> 'Pleiades Sector'
 * ```
 * @example
 * Resolving a permit lock from a position — the exact route, since it does not
 * depend on how the system is named:
 * ```ts
 * import type { GalacticPosition } from '@elite-dangerous-almanac/core/astro/galactic-position';
 * import { findHandAuthoredRegionAt } from '@elite-dangerous-almanac/core/astro/hand-authored-regions';
 *
 * declare const position: GalacticPosition;
 *
 * import { isPermitLockedRegionName } from '@elite-dangerous-almanac/core/astro/permit-locked-regions';
 *
 * const region = findHandAuthoredRegionAt(position);
 * const needsPermit = region !== null && isPermitLockedRegionName(region.name);
 * ```
 */
export function findHandAuthoredRegionAt(position: GalacticPosition): HandAuthoredRegion | null {
    const { x, y, z } = position;
    for (const region of HAND_AUTHORED_REGIONS) {
        for (const s of region.spheres) {
            const dx = x - s.cx;
            const dy = y - s.cy;
            const dz = z - s.cz;
            if (dx * dx + dy * dy + dz * dz <= s.r * s.r) return region;
        }
    }
    return null;
}

File: dotnet/src/EliteDangerousAlmanac/Astronomy/HandAuthoredRegions.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Text.Json.Serialization;
using EliteDangerousAlmanac.Internal;

namespace EliteDangerousAlmanac.Astronomy;

/// <summary>One ball of space a region the game names by hand fills.</summary>
/// <param name="CenterX">The centre's galactic X in light years, with Sol at the origin.</param>
/// <param name="CenterY">The centre's galactic Y in light years.</param>
/// <param name="CenterZ">The centre's galactic Z in light years.</param>
/// <param name="Radius">How far the ball reaches from its centre, in light years.</param>
public sealed record HandAuthoredSphere(
    [property: JsonPropertyName("cx")] double CenterX,
    [property: JsonPropertyName("cy")] double CenterY,
    [property: JsonPropertyName("cz")] double CenterZ,
    [property: JsonPropertyName("r")] double Radius);

/// <summary>A sector the game names by hand, and the space it fills.</summary>
/// <param name="Name">The region name as the game writes it, such as <c>Pleiades Sector</c>.</param>
/// <param name="Spheres">The balls of space the region is the union of.</param>
/// <remarks>
/// Whether the region asks for a permit is not held here. <see cref="PermitLocks"/> is the
/// one place that reads that, and it takes this name.
/// </remarks>
public sealed record HandAuthoredRegion(string Name, IReadOnlyList<HandAuthoredSphere> Spheres);

/// <summary>The sectors the game names by hand, found by where a system sits.</summary>
/// <remarks>
/// <para>
/// The generator only ever writes the boxel name of a place. A system that physically
/// falls inside a nebula or a cluster the game names by hand carries that region's name
/// instead. Each such region is one ball of space or several, in light years with Sol at
/// the origin, and a system belongs to the first region whose space holds it.
/// </para>
/// <para>
/// The catalogue is held smallest ball first, which is how the game settles an overlap:
/// the smallest region that holds the place wins.
/// </para>
/// <para>
/// The word region carries three meanings in this galaxy. This one is a sector the game
/// names by hand. <see cref="SectorName"/> reads a sector the generator names, and
/// <see cref="CodexRegions"/> reads one of the 42 zones the codex records a find against.
/// </para>
/// </remarks>
public static class HandAuthoredRegions
{
    private static readonly Lazy<IReadOnlyList<HandAuthoredRegion>> Catalogue = new(Build);

    /// <summary>Every sector the game names by hand, smallest ball of space first.</summary>
    public static IReadOnlyList<HandAuthoredRegion> All => Catalogue.Value;

    /// <summary>Reads the region one place in the galaxy belongs to.</summary>
    /// <param name="position">The position in light years, with Sol at the origin.</param>
    /// <returns>
    /// The region that holds the place, or <see langword="null"/> where the generator names
    /// it instead.
    /// </returns>
    /// <exception cref="ArgumentNullException">The position is absent.</exception>
    /// <remarks>
    /// All three axes count. A region the game names by hand is a ball of space, so the
    /// height matters here, which it does not for a codex region.
    /// </remarks>
    public static HandAuthoredRegion? FindAt(GalacticPosition position)
    {
        if (position is null) throw new ArgumentNullException(nameof(position));

        foreach (HandAuthoredRegion region in All)
        {
            foreach (HandAuthoredSphere sphere in region.Spheres)
            {
                double dx = position.X - sphere.CenterX;
                double dy = position.Y - sphere.CenterY;
                double dz = position.Z - sphere.CenterZ;
                if ((dx * dx) + (dy * dy) + (dz * dz) <= sphere.Radius * sphere.Radius)
                {
                    return region;
                }
            }
        }

        return null;
    }

    /// <summary>Reads the catalogue, and closes the balls of each region to a change.</summary>
    private static ReadOnlyCollection<HandAuthoredRegion> Build()
    {
        IReadOnlyList<HandAuthoredRegion> records =
            SharedData.LoadList<HandAuthoredRegion>("data/astro/hand-authored-regions.jsonc");
        List<HandAuthoredRegion> regions = new(records.Count);
        foreach (HandAuthoredRegion region in records)
        {
            regions.Add(region with { Spheres = ReadOnlyLists.Freeze(region.Spheres) });
        }

        return new ReadOnlyCollection<HandAuthoredRegion>(regions);
    }
}
